Add C API guides (#2796)
Qiskit/documentation: docs/guides/install-c-api.mdx

Qiskit Python workflow with C, read [Extend Python with the Qiskit C API](/guides/c-extension-for-python).

The following example builds an observable with C:
```c
// file: example.c
#include <stdio.h>
#include <stdint.h>
#include <complex.h>
#include <qiskit.h>

int main(int argc, char* argv[]) {
    // build a 100-qubit empty observable
    u_int32_t num_qubits = 100;
    QkObs *obs = qk_obs_zero(num_qubits);

    // add the term 2 * (X0 Y1 Z2) to the observable
    complex double coeff = 2;  // on MSVC use: _Dcomplex coeff = {2, 0}
    QkBitTerm bit_terms[3] = {QkBitTerm_X, QkBitTerm_Y, QkBitTerm_Z};  // bit terms: X Y Z
    uint32_t indices[3] = {0, 1, 2};  // indices: 0 1 2
    QkObsTerm term = {coeff, 3, bit_terms, indices, num_qubits};
    qk_obs_add_term(obs, &term);  // append the term

    // print some properties and the observable itself
    printf("num_qubits: %i\n", qk_obs_num_qubits(obs));
    printf("num_terms: %lu\n", qk_obs_num_terms(obs));
    printf("observable: %s\n", qk_obs_str(obs));

    // free the memory allocated for the observable
    qk_obs_free(obs);

    return 0;
}
```

> ## UNIX

Reviewer: I'd probably call this `UNIX-like` as neither Linux or macOS is actually Unix itself.

```suggestion
## UNIX-like
```

If we supported IBM AIX or Solaris (which we thankfully don't) we could call this UNIX since those are proper UNIX operating systems.

> This section provides build instructions for UNIX systems.

Reviewer: ```suggestion
This section provides build instructions for UNIX-like systems.
```

> ### Requirements

Reviewer: You might want to point to the platform support page here so people know the platform requirements for building from source.

Compilation requires the following tools:

> * A Rust compiler: see for example the [guide on installing Qiskit from source](https://github.com/Qiskit/qiskit/blob/main/CONTRIBUTING.md#installing-qiskit-from-source)

Reviewer 2: ```suggestion
* A Rust compiler: see for example the [guide on installing Qiskit from source](/guides/install-qiskit-source)
```

* A C compiler: for example, GCC on Linux and Clang on MacOS
* [`cbindgen`](https://github.com/mozilla/cbindgen): a tool to crate the C header, which you can install with `cargo install cbindgen`
    Note that running the tool from the command line should be enabled, which might require exporting your `PATH` variable to include `/path/to/.cargo/bin`

> * Make: this is optional but is recommended to use automated install processes

Reviewer 3: We use GNU extensions in the Makefile, iirc, so specifically you need GNU Make.

---

> title: Installing Qiskit's C API
> description: How to install and use Qiskit's C API as standalone library
> ---
> 
> # Installing Qiskit's C API
> 
> This guide describes how to install and use Qiskit's C API. To see how to extend your
> Qiskit Python workflow with C, check the guide on [Extending Python with Qiskit's C API](./c-extension-for-python).

Reviewer 2: ```suggestion
title: Install the Qiskit C API
description: How to install and use the Qiskit C API as standalone library
---

# Install the Qiskit C API

This guide describes how to install and use the Qiskit C API. To see how to extend your
Qiskit Python workflow with C, read [Extend Python with the Qiskit C API](/guides/c-extension-for-python).
```

> As example program we consider building an observable with C:

Reviewer 2: ```suggestion
The following example builds an observable with C:
```

```c
// file: example.c
#include <stdio.h>
#include <stdint.h>
#include <complex.h>
#include <qiskit.h>

int main(int argc, char* argv[]) {
    // build a 100-qubit empty observable
    u_int32_t num_qubits = 100;
    QkObs *obs = qk_obs_zero(num_qubits);

    // add the term 2 * (X0 Y1 Z2) to the observable
    complex double coeff = 2;  // on MSVC use: _Dcomplex coeff = {2, 0}
    QkBitTerm bit_terms[3] = {QkBitTerm_X, QkBitTerm_Y, QkBitTerm_Z};  // bit terms: X Y Z
    uint32_t indices[3] = {0, 1, 2};  // indices: 0 1 2
    QkObsTerm term = {coeff, 3, bit_terms, indices, num_qubits};
    qk_obs_add_term(obs, &term);  // append the term

    // print some properties and the observable itself
    printf("num_qubits: %i\n", qk_obs_num_qubits(obs));
    printf("num_terms: %lu\n", qk_obs_num_terms(obs));
    printf("observable: %s\n", qk_obs_str(obs));

    // free the memory allocated for the observable
    qk_obs_free(obs);

    return 0;
}
```

> ## Unix
> 
> In the following we provide build instructions for Unix systems.

Reviewer 2: ```suggestion
## UNIX

This section provides build instructions for UNIX systems.
```

### Requirements

We require the following tools for the compilation:

> * A Rust compiler: see e.g. the [guide on installing Qiskit from source](https://github.com/Qiskit/qiskit/blob/main/CONTRIBUTING.md#installing-qiskit-from-source).

Reviewer 2: Any reason not to point to the guide [here](https://quantum.cloud.ibm.com/docs/en/guides/install-qiskit-source) instead? (relative path would be /guides/install-qiskit-source)

> We require the following tools for the compilation:
> * A Rust compiler: see e.g. the [guide on installing Qiskit from source](https://github.com/Qiskit/qiskit/blob/main/CONTRIBUTING.md#installing-qiskit-from-source).
> * A C compiler: we'll use GCC on Linux and Clang on MacOS
> * [`cbindgen`](https://github.com/mozilla/cbindgen): a tool to crate the C header, which you can e.g. install via `cargo install cbindgen`.
>     Note that running the tool from the command line should be enabled, which might require exporting your `PATH` variable
>     to include `/path/to/.cargo/bin`.

Reviewer 2: ```suggestion
Compilation requires the following tools:
* A Rust compiler: see for example the [guide on installing Qiskit from source](https://github.com/Qiskit/qiskit/blob/main/CONTRIBUTING.md#installing-qiskit-from-source)
* A C compiler: for example, GCC on Linux and Clang on MacOS
* [`cbindgen`](https://github.com/mozilla/cbindgen): a tool to crate the C header, which you can install with `cargo install cbindgen`
    Note that running the tool from the command line should be enabled, which might require exporting your `PATH` variable to include `/path/to/.cargo/bin`
```

> * Make: this is optional, but we recommend the installation to use automated install processes

Reviewer 2: ```suggestion
* Make: this is optional but is recommended to use automated install processes
```

> You can check everything is installed via

Reviewer 2: ```suggestion
This code verifies that everything has been installed:
```

```bash
rustc --version
gcc --version
cbindgen --version
make --version  # optional, but recommended
```

> ### Building

Reviewer 2: ```suggestion
### Build
```

To build the C header and library, you can run the following Make command[^1] in Qiskit root,

> [^1]: If you did not install Make, check the `Makefile` in Qiskit root for the required commands. (Or just install Make, it's not too late.)

Reviewer 2: ```suggestion
[^1]: If you did not install Make, check the `Makefile` in Qiskit root for the required commands - or simply install Make; it's not too late.)
```

```bash
make c
```

> which will provide the compiled shared library in `dist/c/lib` and the `qiskit.h` header with
> all function declarations in `dist/c/include`. Note that the precise library name depends
> on the platform, e.g `libqiskit.so` on Unix and `libqiskit.dylib` on MacOS.
> (Note that this step currently emits a lot of warnings. This is fine, we're working on it.)

Reviewer 2: ```suggestion
which will provide the compiled shared library in `dist/c/lib` and the `qiskit.h` header with
all function declarations in `dist/c/include`. Note that the precise library name depends
on the platform; for example,`libqiskit.so` on UNIX and `libqiskit.dylib` on MacOS.
(Note that this step currently emits a lot of warnings, which is expected, and is no cause for alarm. Future versions will remove the warnings.)
```

> We can then compile a C program using the Qiskit C header and library, by

Reviewer 2: ```suggestion
You can then compile a C program using the Qiskit C header and library, with
```

```bash
gcc example.c -o example.o -I /path/to/dist/c/include -L /path/to/dist/c/lib -lqiskit
```
To ensure the library is found during execution, set the runtime library path to

> include `/path/to/dist/c/lib`. This depends on the platform, e.g. on Linux

Reviewer 2: ```suggestion
include `/path/to/dist/c/lib`. This depends on the platform. On Linux:
```

```bash
export LD_LIBRARY_PATH=/path/to/dist/c/lib:$LD_LIBRARY_PATH
```

> or on MacOS

Reviewer 2: ```suggestion
On MacOS:
```

```bash
export DYLD_LIBRARY_PATH=/path/to/dist/c/lib:$DYLD_LIBRARY_PATH
```

> Alternatively, we can set the runtime library path during compilation by adding

Reviewer 2: ```suggestion
Alternatively, you can set the runtime library path during compilation by adding
```

```
-Wl,-rpath,/path/to/dist/c/lib
```
to the compiler flags.

> Now we can execute the binary,

Reviewer 2: ```suggestion
Now you can execute the binary:
```

```bash
./example.o
```

> which, if using the example snippet shown above, should print

Reviewer 2: ```suggestion
which, if using the example snippet shown previously, should print
```

```
num_qubits: 100
num_terms: 1
observable: SparseObservable { num_qubits: 100, coeffs: [Complex { re: 2.0, im: 0.0 }], bit_terms: [X, Y, Z], indices: [0, 1, 2], boundaries: [0, 3] }
```

## Windows (experimental)

<Admonition type="caution">

>   Windows support is currently experimental. Instructions are work in progress.

Reviewer 2: ```suggestion
  Windows support is currently experimental. Instructions are a work in progress.
```

</Admonition>

### Requirements

We require the following tools for the compilation:
* A Rust compiler: see e.g. the [guide on installing Qiskit from source](https://github.com/Qiskit/qiskit/blob/main/CONTRIBUTING.md#installing-qiskit-from-source).
* A C compiler: we'll use MSVC and the native command prompt offering the ``cl`` command
* A Python installation, both with access to ``python3.lib`` and ``python3.dll``.
* [`cbindgen`](https://github.com/mozilla/cbindgen): a tool to crate the C header, which you can e.g. install via `cargo install cbindgen`.
    Note that running the tool from the command line should be enabled, which might require exporting your `PATH` variable
    to include `/path/to/.cargo/bin`.

> ### Building
> 
> First, we compile the `qiskit_cext` dynamic library, by running the following in Qiskit root

Reviewer 2: ```suggestion
### Build

First, compile the `qiskit_cext` dynamic library, by running the following in Qiskit root
```

```bash
set PATH="\path\to\pythonlib\";%PATH%
cargo rustc --release --crate-type cdylib -p qiskit-cext
```

> This will generate the ``.dll`` dynamic library, and associated ``.dll.lib`` file, in `target/release`.
> Next, we have to generate the header via

Reviewer 2: ```suggestion
This will generate the ``.dll`` dynamic library and associated ``.dll.lib`` file in `target/release`.
Next, generate the header with
```

```bash
cbindgen --crate qiskit-cext --output target\release\qiskit.h
```

> which will write the MSVC-compatible header in the same directory as the library.

Reviewer 2: ```suggestion
This will write the MSVC-compatible header in the same directory as the library.
```

> Now we can use `cl` to compile the C program. For simplicity we assume the file is

Reviewer 2: ```suggestion
Now you can use `cl` to compile the C program. For simplicity, here it is assumed the file is
```

located in `target/release`.
```bash
cl example.c qiskit_cext.dll.lib -I.
```

> If your source file has another location ensure to update the

Reviewer 2: ```suggestion
If your source file has another location,  be sure to update the
```
